### ISSUE TITLE
`npm install` after `npm link` will "steal" dependencies from linked packages

### ISSUE THREAD

**Issue by jhenninger** [npm/npm#10343]
Create two packages:

``` sh
jo@t420:/tmp/npmtest$ mkdir A B
```

A depends on lodash:

``` sh
jo@t420:/tmp/npmtest$ cd A
jo@t420:/tmp/npmtest/A$ npm init --yes && npm install --save lodash
Wrote to /tmp/npmtest/A/package.json:

{
  "name": "A",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}


A@1.0.0 /tmp/npmtest/A
`-- lodash@3.10.1 

npm WARN EPACKAGEJSON A@1.0.0 No description
npm WARN EPACKAGEJSON A@1.0.0 No repository field.
```

B depends on A which is symlinked for development:

``` sh
jo@t420:/tmp/npmtest/A$ cd ../B
jo@t420:/tmp/npmtest/B$ npm init --yes                       
Wrote to /tmp/npmtest/B/package.json:

{
  "name": "B",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
```

Create the link:

``` sh
jo@t420:/tmp/npmtest/B$ npm link ../A
npm WARN EPACKAGEJSON A@1.0.0 No description
npm WARN EPACKAGEJSON A@1.0.0 No repository field.
/home/jo/code/nvm/versions/node/v4.2.1/lib/node_modules/A -> /tmp/npmtest/A
/tmp/npmtest/B/node_modules/A -> /home/jo/code/nvm/versions/node/v4.2.1/lib/node_modules/A -> /tmp/npmtest/A
```

So far so good, A still has lodash in its node_modules:

``` sh
jo@t420:/tmp/npmtest/B$ tree -d ..
..
├── A
│   └── node_modules
│       └── lodash
│           ├── array
│           ├── chain
│           ├── collection
│           ├── date
│           ├── function
│           ├── internal
│           ├── lang
│           ├── math
│           ├── number
│           ├── object
│           ├── string
│           └── utility
└── B
    └── node_modules
        └── A -> /home/jo/code/nvm/versions/node/v4.2.1/lib/node_modules/A

18 directories
```

Now actually save A as a dependency of B in package.json:

``` sh
jo@t420:/tmp/npmtest/B$ npm install --save ../A
npm WARN update-linked node_modules/A needs updating to 1.0.0 from 1.0.0 but we can't, as it's a symlink
lodash@3.10.1 node_modules/A/node_modules/lodash -> node_modules/lodash
npm WARN EPACKAGEJSON B@1.0.0 No description
npm WARN EPACKAGEJSON B@1.0.0 No repository field.
```

And all dependencies of A get moved to B's node_modules folder:

``` sh
jo@t420:/tmp/npmtest/B$ tree -d ..
..
├── A
└── B
    └── node_modules
        ├── A -> /home/jo/code/nvm/versions/node/v4.2.1/lib/node_modules/A
        └── lodash
            ├── array
            ├── chain
            ├── collection
            ├── date
            ├── function
            ├── internal
            ├── lang
            ├── math
            ├── number
            ├── object
            ├── string
            └── utility

17 directories
```

Running tests etc. in A will now fail due to missing dependencies, unless you run
`npm install` in A again.

This will also happen when manually adding A as a dependency to B's package.json
 and then running `npm install`.

npm version: 3.3.12

**Comment by iarna** [comment]
So, I think this is caused by two things:

1. https://github.com/npm/npm/commit/33aedafd9531b897dca6f650ed09aab483ca3163 which caused npm to move stuff around when it shouldn't.

AND

2. A specific bug that allows moves to happen from inside linked modules. (This https://github.com/npm/npm/blob/v3.4.0/lib/install/diff-trees.js#L145-L152 should not generate moves if the source is inside a link. But further, https://github.com/npm/npm/blob/v3.4.0/lib/install/diff-trees.js#L119-L126 probably shouldn't generate removes if the source is inside a link. Possibly these should be like update-link, which is to say, a different action that just warns.)

**Comment by baelter** [comment]
+1 this is burning time = money and sanity. Target release? Need help?

**Comment by ElliotChong** [comment]
Ghetto hacky workaround idea that might help the next developer: Try using the NODE_PATH env variable to treat `B` as a `node_modules` path.

`NODE_PATH=/path/to/B/node_modules node index.js`

https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders

**Comment by timdp** [comment]
I've had _some_ success working around this issue by rebuilding my node_modules folder with [this script](https://gist.github.com/timdp/1ffc2af2c53ffd3591d0) (written for Windows). However, it often triggers "refusing to delete: containing path isn't under npm's control" errors. I'm sure that's a different issue underneath. Regardless, `npm link` is just broken at this point.

**Comment by whitecolor** [comment]
+1 when it is going to be fixed?

**Comment by tehsenaus** [comment]
Any update on this?

**Comment by whitecolor** [comment]
That actually want holding many from using npm3.

**Comment by baelter** [comment]
@whitecolor Maybe not but it's still an undesirable "feature"

**Comment by baelter** [comment]
I think the whole link command seems a bit over complicated maybe? Correct me if I'm wrong but wouldn't the following do what we want?

```npm link x``` should only do:
1. force move node_modules/x to node_modules/.x
2. create symlink

```npm unlink [x]``` should only do:
1. If no argument, do for all symlinks, else do for x:
2. remove symlink
3. move node_modules/.x to node_modules/x

The you should probably carry on with an ```npm update```, but that should maybe not be a part of linking.

Ps. We should also have a ```npm list links```

**Comment by larsenjh** [comment]
+1 please fix this. We have a very complex module structure, and this problem causes many of us to have to re-install all our deps in all our projects over again. Terrible bug.

**Comment by ecdeveloper** [comment]
+1

**Comment by timdp** [comment]
@baelter As I understand it, `npm link` itself is fine, but `npm install` doesn't treat links any differently (or maybe it does, but badly), which causes inconsistencies.

**Comment by baelter** [comment]
@timdp It looks to me like ```npm link x``` does a lot more that what I suggested above. ```npm link``` is simple enough and I haven't found any issues with that.

**Comment by schmod** [comment]
@timdp Agreed, and I'm not sure whether or not this is intentional.  If nothing else, it's a significant departure from npm2.x's behavior.  

For example, if I have `smallApp` that depends on `bigLib`, I might do something like this:

```sh
$ cd bigLib
$ npm install
$ npm link

$ cd ../smallApp
$ npm link bigLib
   ~/smallApp/node_modules/bigLib -> ~/npm/lib/node_modules/bigLib -> ~/bigLib
$ npm install 
```

If `smallApp` has no other dependencies, I would expect that final `npm install` command to do nothing, because all of `bigLib`'s dependencies are already installed in `bigLib/node_modules`.  (This what npm2.x would have done)

Instead, running `npm install` from `smallApp` installs a duplicate set of `bigLib`'s dependencies into `smallApp`'s `node_modules` folder.  Because `bigLib` is big and has many dependencies, this could potentially be a very lengthy operation.

---

Ideally, npm3 should avoid tree-flattening when traversing into linked packages.  I'm having a difficult time envisioning a situation where the current behavior is desirable.  

---

One tricky case worth considering: 

If `bigLib` is missing a dependency, should npm install a copy of the missing dep into `smallApp`? 

Or can we enforce a firm contract that `npm` does not resolve transitive dependencies for linked modules, thus making the linked module completely responsible for providing its dependencies?)

**Comment by i-e-b** [comment]
@schmod 

> thus making the linked module completely responsible for providing its dependencies?

I would say this is the most sensible option, both closest to npm 2, and it supports the case where a linked dependency may be part of a wider project and is being included in various places for packaging reasons.

**Comment by timdp** [comment]
I'd just always install/keep all the linked package's dependencies inside its node_modules, regardless of whether you ran npm install for the package itself or for one that links to it. That way, at worst, you'll end up with some duplicate dependencies caused by changes to package.json, which should be easy to fix with npm prune. At the same time, all your packages will have all their dependencies available, no matter how they're getting loaded. But I sense I'm missing something, since I don't understand why this continues to be an issue in npm 3.

**Comment by 1mjwebbthew** [comment]
endless frustration with this issue, endless.

**Comment by schmod** [comment]
I have a PR open that appears to address this issue under the following circumstances:

* Module B has a symlinked copy of Module A in its `node_modules` folder.  (Possibly, but not necessarily created with `npm link`)
* Module B's `package.json` lists Module A as a dependency or devDependency
* This dependency is listed **without** a version specifier (ie. it is a FS or Git dependency)

Has anybody been able to reproduce this issue _outside_ of those constraints?  As far as I've been able to tell, this bug is limited to non-repo dependencies.

**Comment by tellnes** [comment]
@schmod Yes, I'm seeing this issue for dependencies with version ranges. And I've not jet seen a consistent behavior for when it is stealing and when it is not.

The solutions is usually to rerun `npm install` a few times in the module that is loosing dependencies after I've run `npm install` in the module that is stealing them. But this fast gets inconvenient when you have a tree of dependecies that is linked together.

**Comment by schmod** [comment]
@tellnes Do you have a simple scenario/repo that can reliably reproduce that behavior?

I spent quite some time trying to reproduce this bug with version ranges, and was unable to do so.

**Comment by tellnes** [comment]
@schmod I'll see next week if I can reproduce this in a single setup.

Most of the modules are private, hosted with Sinopia. My current setup have multiple levels of symlinks down in the dependency tree.

**Comment by iarna** [comment]
Just FYI, this is my current priority. This does not seem to be resolved by the fix for #10800, which is quite interesting.

I believe the original scenario should have _removed_ the link to A and installed a copy of it instead, as that's what `npm install --save ../A` means, I think what was intended was `npm install --save --link ../A` but I see that doesn't work _either. =/

But from what you all are saying I think the problem you all are typically having is more along the lines of what #10800 is describing, and that, that I have a fix for and a (just now) written super minimal reproducer test, so I expect that to land next week.

**Comment by schmod** [comment]
@iarna If you haven't seen it, I have [a PR](https://github.com/npm/npm/pull/12828) that mitigates this somewhat, although I hadn't considered the impacts of using `save` on a linked package with my fix...

**Comment by iarna** [comment]
@schmod: I just took a looksey, I'll followup over there.

**Comment by ggarek** [comment]
There is one more thing to consider.

*It seems that `npm link` must be more sophisticated then simply a symbolic link.*

**Use case:**
You develop project A and project B. Project B is a library that will be published to npm. For development purposes you use `npm link` to work on B.
The B library is used in A, so you do `npm link B` in A directory.

As for now *(npm version 3.10.7)* the B appears under `A/node_modules` as a symbolic link to B working directory.

**What is wrong with this?**
The problem is that, `B/node_modules` may contain dev dependencies, which in turn may “shadow” dependencies in A.

**Possible solution:**
As i see it, the `npm link` must somehow provide the same result as `npm install` does. 
*In other words, when you link a library, it must not contain local node_modules directory, and all dependencies must be installed flattened in the main(root) project.*

**How did i found the case?**
I have two projects: main and library. I bundle the main project with webpack. And when i develop my library locally with `npm link` i end up with duplicate packages in the resulting bundle.
So, may be it is webpack`s problem, but i do not see it this way.

**Comment by schmod** [comment]
@ggarek: I'm not sure if I understand the problem.  

If you're depending on B's dependencies to always show up in `A/node_modules`, that's probably not reasonable -- anything that A depends upon needs to be explicitly listed in `A/package.json`.  

On the other hand, `npm link` does seem like it could change the way that conflicts are resolved.   If A has `{ dependencies: "dep1": ~1.1.2}` and B has `{dependencies: "dep1": "~1.1.4"}`, A and B would normally share `A/node_modules/dep1@1.1.4`, but if we `npm link` B, we might get `A/node_modules/dep1@1.1.2` and `B/node_modules/dep1@1.1.4`, which would be perfectly valid, but would cause Webpack to bundle two versions of dep1.

One potential solution for this problem (which may possibly be a justification for npm3's "stealing" behavior) is to consider B's dependencies when deciding what version to install in `A/node_modules`.  Running `npm install` from A could install the most recent package that satisfies both A and B's dependencies, even if B is symlinked.  (This is less than ideal from a performance standpoint, because it requires us to traverse and discover all of B's transitive dependencies, and ends up being a more complicated operation than the non-symlink version)

On the other hand, if dep1@1.1.4 is installed to both `A/node_modules/dep1` and `B/node_modules/dep1`, most users probably won't want Webpack to include two identical copies of dep1 in its bundle.  (This is currently **not** how Webpack handles this situation, but that's not NPM's problem: See https://github.com/webpack/webpack/issues/554)

---

My $0.02 is that Webpack, Browserify, and similar browser bundlers should handle this on their own, or encourage their users to avoid symlinks entirely.  Anything else seems like it leads to confusing behavior.  

I personally don't like a lot of the solutions proposed in https://github.com/webpack/webpack/issues/554 -- I personally work around this problem by using something this in my webpack configuration:

```js
resolve: {
      alias: {
        jquery: require.resolve('jquery'),
        lodash: path.dirname(require.resolve('lodash')), // resolve to a path so we can import things from lodash/fp
      }
}
```

While this requires extra configuration (which I generally dislike), I think that it would be possible for Webpack to detect this situation, and warn users when it's happening.  From there, users can decide if they want to add this workaround to their configuration if they do indeed want Webpack to deduplicate the dependency (and therefore diverge from Node's resolution behavior).

---

If none of these workarounds are acceptable, there really aren't any options that don't involve symlinks.

For this reason, my unhappy conclusion is that production builds and production applications probably need to assume that a package is _always_ symlinked, _never_ symlinked, or acknowledge that the set of packages may slightly differ between those two configurations.

However, many users don't have the ability (or desire) to publish all of their packages to a public NPM repository.  To best accommodate this scenario, NPM would need to implement a Maven-style local repository, or package maintainers should run `npm pack` on their private/local modules, and point `package.json` of the packages consuming those libraries at the resulting tarballs.

**Comment by jerrybendy** [comment]
I had found an issue about `npm link`. I want install some modules global, and link it to my project when use it. Such as `webpack`, `node-sass`, etc. 

I had do that:

```
# install webpack global
sudo npm i -g webpack

cd ~/my-project

npm ln webpack
```

then, it created `./node_modules/webpack` symbol link. 

My `package.json` has a dependencies of `webpack`. 

```json
  "devDependencies": {
    "webpack": "^1.13.2"
  }
```

Now, when I use `npm i` to install my dependencies, I got an error:

```
npm ERR! Error: EACCES: permission denied, access 'node_modules/webpack/node_modules/webpack-core/node_modules/source-map/node_modules'
```

It tell me has no permission, means it will re-install webpack to over-write my symbol link .

What should I do?

**Comment by whitecolor** [comment]
@jerrybendy unlink everything, do npm install , link again, and then install/update modules one by one. 

It still not fixed, and maybe it won't for years to come.

**Comment by jerrybendy** [comment]
@whitecolor   Yes, I just do like you said. 

I removed my dependencies which need to be linked, and run `npm i`. After install complete, do link again.

**Comment by ggarek** [comment]
@schmod Very detailed answer, thank you, you make a good point.

But the issue is not only bundler's "problem", it is about how the `require` algorithm works, so it is the case for pure node apps too.

I think, that using private npm registry is a right way (we use it in my team). But it is good for using published (local registry or public) libraries and the issue is related for development time mostly.

*Regarding my situation in particular:*
The B library depends on a public lib in two ways: as devDependency and as peerDependency. So from within A project, B should resolve proper version/instance of the public lib as of peer dependency. But instead it resolves the public lib as own devDependency. (so the public lib is loaded twice, and the version may differ)

I read  [webpack/webpack#554](https://github.com/webpack/webpack/issues/554) through and i agree, there was no solution proposed, all offers are yet workarounds with obvious flaws.

I am not ready to argue which tool should handle this case, but it is still an issue, may be rare though.

**Comment by dryajov** [comment]
I have a similar problem that @ggarek is describing. Linked modules and the way require works under node breaks singleton functionality. As it has been described above, linked modules with devDependencies overshadow dependencies in the parent module, this is a problem with modules such as mongoose that rely on a singleton as a default mechanism to share connections. 

In my particular case, I'm running into issues with aurelia DI, where the container supports registering singletons to types (such as classes/constructor functions), this breaks if a child module loads its own version of the dependency, since the container will not be able to resolve it to the already registered code reference. As this approach to using DI becomes more wide spread, as well as in the case of singletons, this is critical functionality that breaks under npm link.

Here is an example to illustrate a container that supports registering code references to singletons:

```
// npm module A in file a.js
export class A {
}

// npm module B in file b.js
import {A} from 'a';

@autoinject
class B {
  constructor(a: A) {}
}

// npm module C in file c.js that brings in module B which also brings in module A
import {A} from 'a';

@autoinject
class C {
 constructor(a: A){}
}

// register.js also in npm module C
import {A} from 'a';

function registerDependecies(container: Container) {
  container.registerSingleton(A);
}

registerDependencies(container);
```

When the linked module B gets used in C it brings its own version of A which C also includes directly, this breaks the container functionality since when A is looked up for in the container for B, it gets a different code ref, which ends up resolving to a new instance of A.

**Comment by NiXXeD** [comment]
This has been a pretty big issue for us as well. We've been using npm@2.x for now, but with the news that this version may be only receiving maintenance updates, I am hesitant to stay on it. I just verified this issue today as well.

* local projects Package A depends on Package B.
* `npm i` in Package B directory
* `npm link` in Package B directory
* `npm link packageb` in Package A directory
* `npm i` in Package A directory

Many modules will have been removed from package  b's `node_modules`. You can run `npm i` again in package b to fix it, but when you have multiple levels of this, it gets very tedious and mysterious when modules disappear.

**Comment by shazron** [comment]
We have this issue in Apache Cordova as well. Here's a repro:
```
PACKMAN=npm

mkdir repro-npm-issue-10343-using-$PACKMAN
cd repro-npm-issue-10343-using-$PACKMAN

git clone https://github.com/apache/cordova-js --depth 10
cd cordova-js

$PACKMAN install
$PACKMAN unlink
$PACKMAN link
cd ..

git clone https://github.com/apache/cordova-lib --depth 10
cd cordova-lib

cd cordova-common
$PACKMAN install
$PACKMAN unlink
$PACKMAN link
cd ..

cd cordova-fetch
$PACKMAN link cordova-common
$PACKMAN install
$PACKMAN unlink
$PACKMAN link
cd ..

cd cordova-lib
$PACKMAN link cordova-js
$PACKMAN link cordova-common
$PACKMAN link cordova-fetch
$PACKMAN install
$PACKMAN unlink
$PACKMAN link
cd ..

cd ..

git clone https://github.com/apache/cordova-cli --depth 10
cd cordova-cli
$PACKMAN link cordova-lib
$PACKMAN link cordova-common
$PACKMAN install

npm test
```

The above is a failing test. 

If you set the variable `PACKMAN` in the script above to the alternative package manager [`yarn`](https://github.com/yarnpkg/yarn/) it passes.

**Comment by nmschulte** [comment]
I believe I'm experiencing this issue on Linux now, too.  I've reconfigured my system from the ground up, and that means I've rebuilt node and npm (updated the versions).  I wonder if that's related.  I wasn't far behind before, w/ a v6 node src-tree.

**Comment by admosity** [comment]
Experiencing this issue when doing compilation with sass that has a node_modules include path and the required package is not being installed into node_modules, in my case `foundation-sites` is required by the linked package I am using. I guess the quick workaround would be to include it in my dependencies.

**Comment by doriansmiley** [comment]
We are experiencing this issue as well. If we set up npm link in project a, link project a in project b, then run npm install --save in project b dependencies are then missing in project a.

**Comment by ourmaninamsterdam** [comment]
@doriansmiley ditto

**Comment by brycejacobs** [comment]
This is endless frustration when trying to work with webpack. Has anyone found a decent workaround? A docker dev environment is impossible with this issue.

**Comment by alan-agius4** [comment]
I ended up creating my own 'link' function.

**Comment by metodribic** [comment]
Is there really no solution/workaround for this after more than a year ?

**Comment by magnushiie** [comment]
I've switched to yarn which doesn't have these issues with link.

**Comment by alan-agius4** [comment]
Is this issue still present on npm 4.x.x?

**Comment by ggarek** [comment]
@alan-agius4 yeah, it is still an issue.

I also use yarn for some projects.
For other projects using npm, i just use a task to watch, build and copy built library _(as it would come from npm registry)_ to the node_modules of the dependent project.

**Comment by nmschulte** [comment]
I've started using a shell script which does some manual operations around `npm link`, to avoid the "stealing" issue.  It has the caveat that it completely removes the `node_modules` directory and the global links.  This solution works for me, without having to completely jump ship to e.g. `yarn`.  I can share the script.

**Comment by metodribic** [comment]
Can we somehow help with this issue? It seems that it is not that problematic to fix this since every other here can write his/her own solution/workaround or is there any hidden complexity that I'm not aware of?

**Comment by nmschulte** [comment]
I've sanitized the script, though I didn't bother to make it entirely generic.  The script works to make `npm link` useful for a "complex" (read: actually utilizing `npm link` for something non-trivial) `npm link`ed project.

https://desmas.net/npm-install-link.sh

The project it's used with (`projectName` in the script) has a package for:
1. the "entry point" ("main"; `projectName` repository/package/module), 
2. the "common core" ("core"; `projectName-core` repository/package/module),
3. the "shared UI components" ("ui"; `projectName-ui` ...),
4. independent "application modules" ("appModule`n`", `projectName-appModuleA` ...)

Items 2 and 3 are what cause "trouble," because, not only does 1 link to 2 and 3, but 3 links to 2, and 4 and beyond all link to 2 and 3 as well.  The script works around this trouble by doing things in careful order, and manually installing some links and what not.

The real trick is that, when you need to re-install packages, you must start from scratch and blow away `node_modules`.  This is a _nasty_ work around.  I suspect this could be improved to not need to start from scratch each time, but it's proven to work well so far.

It's probably worth noting that the project I'm using this with uses explicit versioning in `package.json` for dependencies; having "loose" versioning seems to conflate the bug here, as well as causing other problems with a superficial cross-repo/package split of src.

**Comment by mividtim** [comment]
How is it that this issue has been totally ignored by the maintainers for more than a year?  I see no NPM project maintainers talking on the thread, and it's unassigned.  Anyone building a major project in Javascript runs into this issue.  Ignoring a problem because "most users" don't experience it is a bit unprofessional, given that this tool is built into Node itself.

**Comment by alan-agius4** [comment]
What I ended up doing is a custom link method.  If anyone wants it, it's attached. I am not 100% sure it works perfectly cause I had to strip out some code that is not relevant for you. But should be a good starting point.

Script below;
[custom-link.ts.txt](https://github.com/npm/npm/files/949783/custom-link.ts.txt)

to run it you can use [ts-node](https://www.npmjs.com/package/ts-node)

**Comment by theBull** [comment]
I've just managed to fix this by deleting the node_modules folder from the actual package (the initial link package) and then re-ran `npm install`. Seems too wildly obvious of a fix to be "the fix" for such a heated issue, but...for those of you out there who stumble upon this and haven't tried that...try that.

**Comment by MikeMatusz** [comment]
@theBull That does fix it, but if you're making changes in the linked package and want to re-compile and test them in your main app, you need to re-install all of the linked package's dependencies every time.

Now that npm@5 has made linking the default behavior for file references (#15900), they've removed one of the workarounds of actually treating it like an installed package, unless I use a script to copy over everything (in package.json files, *not* node_modules) manually like some other comments here have mentioned.

**Comment by dcaillibaud** [comment]
It's difficult for me to understand why such an issue is still open after 3 versions of npm. since npm2, each version is worse on this purpose, working on several main projects including a lot of npm modules of my own (some of them including others).
Building & pushing every modification in a dependency before reinstall the new version in another dependency (looping until top parent) is really painfull, but redo an `npm i` in each parent then re-link then `npm i` in each dependency (and re-link if needed, so loop again install & link on dependencies) is almost worse :-/
I understand that webpack could duplicate modules, but I know that and I have a staging area to clean this (without links). And anyway, most of modules expose a dist folder builded in es5 for a larger compatibility, so even without links, modules would be duplicated...

Please, add an option or an npm environment where npm install doesn't touch anything in any linked dependency !

**Comment by PierBover** [comment]
When I do `npm install` the symlinks to global packages are removed.

Is this expected behaviour? If not, can anyone from the team explain how to use `npm link` properly?

What I would expect is that if I do `npm link some-package` then `npm install` should never touch that symlink to the global package until I do `npm unlink some-package`.

**Comment by metodribic** [comment]
@iarna is there any particular reason this thread is being ignored for this long?

**Comment by HosseinAgha** [comment]
Is this fixed yet? This is the only issue preventing us from migrating to NPM 5 and Node 8

**Comment by BeekeeperAlex** [comment]
Ugh. I didn't want to have to explore this new-fangled "yarn" but that seems to be where everyone is forced to go. Guess I'll be spending some time learning that today. Why do we need another npm? Why can't npm just work?

**Comment by dkent600** [comment]
I don't understand how `npm link` can be useful for anything except the most trivial of scenarios.

**Comment by BeekeeperAlex** [comment]
@dkent600 Then you clearly haven't written very many libraries in conjunction with other applications. Npm link is the only easy way to test changes during development. If it worked like it's expected to that is.

**Comment by dkent600** [comment]
@chevex  *If* is the keyword there.

**Comment by BeekeeperAlex** [comment]
@dkent600 It sounded like you were saying we shouldn't worry about it getting fixed because `npm link` isn't useful in general. My mistake.

**Comment by dkent600** [comment]
@chevex I would LOVE to have this tool working properly.  I really need it.  I'm wasting a ton of time trying to figure out what is going on and how to work around it.

**Comment by basicdays** [comment]
For what it's worth, I just tried `npm install <folder>`, where `<folder>` is the local package that I actually want to link to. It seems to create the link in the `node_modules` directory correctly, it doesn't hoist all the packages modules out of the project, and `npm install` doesn't screw everything up. Dunno, seems `npm install` pretty much replaces `npm link` for my needs at the moment. ¯\\\_(ツ)\_/¯

**Comment by BeekeeperAlex** [comment]
@basicdays Yes that will put it in your node_modules folder but you need to do that every time you make a change to `<folder>`. You can't simply make a change and then run your app. You need to reinstall it every time.

**Comment by basicdays** [comment]
@chevex That doesn't seem to be what I'm seeing. The actual file that's in the main project's `node_modules` is indeed the sym link that we want to the linked module. The package.json does reflect that it's a local file dependency. When I run `npm install` again later, it correctly doesn't clobber the linked package's `node_modules`. When I make an update to the linked package, the main project seems to automatically get the updated code. Like, this is _exactly_ what `npm link` should be doing, I have no idea why it's `npm install` that does it correctly (on my machine?).

**Comment by BeekeeperAlex** [comment]
Well now I have to try lol. Will report back.

**Comment by basicdays** [comment]
Please do! I really hope it works on more than just my machine. :)

**Comment by BeekeeperAlex** [comment]
This does work. In package.json it adds a dependency with the format `file:<path-to-file>`.
Example: 
```
"dependencies": {
  "my-library": "file:../my-library"
}
```

`npm link` doesn't modify package.json at all which would be preferred in order to preserve the real dependency without accidentally committing this hard reference, but the above does appear to work as expected on OS X. It maintains a symlink to the specified file and re-running `npm install` does not nuke anything in the linked projects' folder. Will try Windows tonight. Thanks for that @basicdays!

**Comment by alan-agius4** [comment]
Is this worksround with npm 5  or earlier versions aswell?

**Comment by BeekeeperAlex** [comment]
@alan-agius4 I tested with the latest version of npm (5.0.3).

**Comment by BeekeeperAlex** [comment]
So I think @basicdays may have discovered the fix they implemented here and this issue was simply left open/forgotten.

I just decided to go back and repro the original issue. First I did it with npm version 4.2.0.

```
$ mkdir A B
$ cd A
$ npm init --yes
$ npm install --save lodash
$ cd ../B
$ npm init --yes
$ npm link ../A
$ npm install --save ../A
```

![npm4-link-issue](https://user-images.githubusercontent.com/874394/27561871-228b1c16-5a87-11e7-992c-4b48e7d23ed2.gif)

This correctly reproduced the issue. Lodash disappeared from A's node_modules and appeared in B's node_modules. Notice that the reproduction steps link to the module and *then* install the module from the local file system.

I verified that in npm 4 it did indeed **copy** the module from the referenced path into B's node_modules folder. It did *not* create a symlink, hence the need for `npm link` in addition to local file system `npm install`.

I then went forward and tried the reproduction steps in npm@5 and got some interesting results.

![npm5-link-issue](https://user-images.githubusercontent.com/874394/27562624-858e1860-5a8a-11e7-9bf6-8dd764580ea1.gif)

Notice that when I did `npm install ../A` portion after doing `npm link ../A` it just nuked A's node_modules folder. I'm not sure why that happened but B's package.json file did get updated with a reference to A on the file system so I decided to just run `npm install` after that. Suddenly the tree looked normal with lodash living in A's node_modules folder as expected.

What I am taking away from the above is that `npm link` felt redundant because you can just do this:

![npm5-link-issue-solved](https://user-images.githubusercontent.com/874394/27562857-cc549dfe-5a8b-11e7-8323-8c3f70f4fd2c.gif)

One thing I've yet to do is test `npm link` with the package installed from the registry instead of the file system. I'll do that next. Installing from file system now seems to create a symlink whereas in npm@4 it actually copied the files, making it a better simulation of installing from the npm registry. Since it now creates a symlink the original reproduction steps don't apply. For now it looks like @basicdays has figured out a hell of a workaround (or possible official solution that nobody has updated us about). I'll keep testing around but this behavior hints to me that it's intentional.

**Comment by BeekeeperAlex** [comment]
I can't reproduce the original bug of the linked project's dependencies being flattened into the parent project's dependencies with npm@5, not even with a library that is on the registry. So the original bug isn't an issue anymore. Here I'll go through the scenarios I set up and why I think this new behavior is intentional. Whether or not we users should be happy with it or not is up for discussion.

Here are the following scenarios I've tested with a module that lives on the registry and on my machine somewhere (remember that `--save` is implied in npm@5+):

> **Just a quick note for anyone confused about `npm link`:**
> 
> `cd app && npm link path/to/dependency`
> is equivalent to 
> `cd path/to/dependency && npm link`
> followed by
> `cd app && npm link dependency-name`
> 
> Both ways create a symlink from the dependency to npm's global node_modules and then a symlink from the app to that global symlink.

## Creating link after install.

```
$ mkdir app && cd app
$ npm init -y
$ npm install registry-package
$ npm link local/path/to/registry-package
```

In this scenario the installed registry package is overwritten by the symlink. Any installed modules in the local symlinked package seem untouched 🎉 

## Installing package after linking it.

```
$ mkdir app && cd app
$ npm init -y
$ npm link local/path/to/registry-package
```

This so far correctly links to the local package and that package's dependencies are untouched 🎉 

But then if you:
```
$ npm install registry-package
```

It will overwrite the symlink with the actual package from the registry. However there is a weird side effect here. The previously linked package who's symlink just got overwritten also gets its node_modules directory nuked from existence for seemingly no reason 😠

What I am taking away from this is that `npm link` no longer seems intended to be used to create a permanent local link. It's more for a quick link for debugging. For example, if you're having problems in your app and you've traced it to a dependency then you can quickly `npm link path/to/dependency` and you're good to debug real quick without affecting your app's package.json reference to the actual registry package.

If you want a permanent link to a local package, say for an app still in active development, you would run `npm install path/to/dependency` which permanently symlinks the package into your app with a `"dependency": "file:path/to/dependency"` reference in your package.json. Then when you're ready to go live you would publish your dependencies and replace those references in your app with the actual registry packages.

This is all poorly documented and nobody has stopped by this issue to update us, but this behavior seems deliberate to me. I think this particular issue can be closed. Though I might want to open a new issue for that odd side effect where installing over an existing symlink gives the symlinked package a courtesy wipe of its node_modules for no reason.

**Comment by dcaillibaud** [comment]
Here is an history of this bug.
- npm2 : all is fine and work as expected (npm link create symlink and npm install doesn't change anything in linked packages)
- npm3 & npm4 : npm link create symlink, but npm install move dependencies of the linked package under the main project =>  linked package himself is broken, you need to re-run npm install in it
- npm5 : same bug as npm3 & 4, but it's worse as npm install destroy all symlinks

My need is to work with a modular approach, here is a sample hierarchy
- projectA
-- myModule1
--- myModule2
-- myModule3
--- myModule2
-- projectB (can be run as standalone app or embedded app)
--- myModule4

- projectC
-- myModule1
--- myModule2
-- myModule4

with npm2, npm link create symlink and I can have all my code at only one place on my disk, so during development changing something in a project dependency change it everywhere, if it broke something I see it immediatly (as long as a test exists and is running in background), without pushing it on a repo increasing version, re-install it on another project changing the project package.json and seeing then something missing.

with npm3 & 4, I need to re-run npm install in all my own dependencies after doing it in a project.

npm5 is clearly unusable in this case.

I don't want to use lib approach with link in the package.json, because I want each module to have it's own repo, and no link in production.

**Comment by BeekeeperAlex** [comment]
I'm not saying it's right or wrong the way it is now, especially since all we have is radio silence. But I'm not sure I understand why each module can't have it's own repo in the setup I demonstrated above? Personally I kind of like that I can `npm install ./path/to/myModule` and get a permanent symlink because it allows me to nuke my app's node_modules and run `npm install` and have full confidence my symlinks will return, with their own dependencies intact as well. In npm 4 and lower running `npm install ./path/to/myModule` did not create a symlink but it does in 5.

My only worry is that I might accidentally commit that reference one day and break my app's build. If that's all you mean then I agree that's a potential nuisance. I'd almost prefer to have a separate file that gets used as an override during `npm install` and that file could contain the symlink references but be ignored in source control.

I do however see potential with this if I were to say, swap those references out for the real thing in a build script. Then in dev I could have permanent symlinks that I never have to restore, even if I totally delete node_modules in my app. It seems to me that hierarchy you described is definitely doable in npm@5 with the only downside being that your package.json would have `"yourModule": "file:./path/to/yourModule"` and you'd have to replace those with actual registry references before going to staging/production.

For example, you could potentially create something like a `sym-map.json` file as a manifest of what to swap out when it's time to go live.

```javascript
// projectA/package.json

{
  ...
  "dependencies": {
    "myModule1": "file:../myModule1",
    "myModule3": "file:../myModule3",
    "projectB": "file:../projectB"
  }
}
```

```javascript
// projectA/sym-map.json

{
  "myModule1": "^3.9.1",
  "myModule3": "^1.2.1",
  "projectB": "^4.0.3"
}
```

Then just have your build script reference that file during build and swap them out.

```javascript
var manifest = require('package.json');
var symMap = require('sym-map.json');
var fs = require('fs');

Object.keys(symMap).forEach((name) => {
  manifest.dependencies[name] = symMap[name];
});

fs.writeFile('./package.json', JSON.stringify(manifest));
```

You could even turn that into a script you could run at any time to swap back and forth.

```javascript
// swap-references.js

var manifest = require('package.json');
var symMap = require('sym-map.json');
var fs = require('fs');

Object.keys(symMap).forEach((name) => {
  let manifestValue = manifest.dependencies[name];
  manifest.dependencies[name] = symMap[name];
  symMap[name] = manifestValue;
});

fs.writeFile('./package.json', JSON.stringify(manifest));
fs.writeFile('./sym-map.json', JSON.stringify(symMap));
```

Then run `node swap-references.js` to switch back and forth at will.

It's a little more work but I think the package.json reference might actually provide more utility by making your symlinks more permanent while in dev. It's more explicit rather than expecting `npm install` to detect and skip over symlinked modules in node_modules. Not to mention that in the old way if you nuked node_modules in your app then you'd have to do the `npm link yourModule` all over again.

Anyway, just an idea. Not saying you're wrong at all. I just want to understand your needs better to see if there is indeed no way to accomplish what you're after with npm@5.

**Comment by laat** [comment]
**EDIT**: Nevermind, I cannot make this work with deeper symlink chains `pkg1 -> pkg2 -> pkg3`
Ignore this comment.

This ~seems to work~ in my tests, as it is not breaking symlinks or steals dependencies
```
$ npm -v
5.0.4

$ npm link other
$ npm install --preserve-symlinks --no-shrinkwrap
```

But, as in the original bug report this still steals dependencies:
(the flags do not seem to matter)
```
$ npm link other
$ npm install --preserve-symlinks --no-shrinkwrap ../other
```

I would love to see this issue fixed, preserving the symlink behaviour from npm@2 on `npm install`

**Comment by BeekeeperAlex** [comment]
@laat there's a few issues with your examples.

1. Those flags aren't causing the difference between your two scenarios (and I'm pretty sure npm has no `--preserve-symlinks` flag as that is a flag you pass to node, not npm.)

2. `npm link` does not modify package.json. So running `npm link other` and then `npm install` has no conflicts because `npm link` didn't add anything to package.json and `npm install` by itself just reads from package.json to perform installs.

3. Your second scenario is exactly what I described above with npm 5. If you do `npm link other` and then `npm install ../other` the symlink created by `npm link other` gets overwritten by a symlink created by `npm install ../other`. Npm 4 and earlier did not create symlinks if you did `npm install ../some/path` but in npm 5 it does create one, making `npm link` redundant. The difference is that `npm install ../other` does modify package.json and adds this explicit reference:

    ```
    "dependencies": {
      "other": "file:../other"
    }
    ```

    With that reference in package.json `npm install` will now always create a symlink to that file path, even if you delete `node_modules` altogether and run `npm install` fresh.

4. Your second scenario doesn't actually "steal" dependencies from "other". There's an odd side effect when you use `npm install ../other` to overwrite a symlink created by `npm link other` where it straight up deletes the `node_modules` directory from "other".

Basically I believe they've changed how they expect us to use `npm link` and just haven't updated us here or the documentation.

- If you want a temporary link to "other" then you run `npm link other` and overwrite the installed module with a symlink. Running `npm install` after that will overwrite the symlink with the actual package from the registry again. `npm link` is not intended to create a permanent link anymore.

- If you want a permanent link to "other" then you run *only* `npm install ../other` without using `npm link` at all. It adds a reference to package.json and creates a permanent symlink to "other" that won't be overwritten unless you change the reference in package.json to point to the registry. This allows you to run `npm install` as much as you want and it won't affect that permanent link to "other". To change it to point to an actual package you would run `npm install other` changing package.json from `"other": "file:../other"` to the traditional format `"other": "^1.2.3"`. which points to the registry.

**Comment by dcaillibaud** [comment]
1) I want permanent link in my dev env, because my apps and their dependencies are in a living stage, not just today. having to run `npm link` on every dependency of my own is really painfull, and lead to a lot of duplicates (dependencies of dependencies, placed at the root level by the npm install)
2) I don't want to put `"dependency": "file:path/to/dependency"` in my package.json because path/to/dependency is specific on my local machine, others people coding on this app have others path (or none, if they don't touch dependencies code), and in production I don't want any link.

**Comment by BeekeeperAlex** [comment]
@dcaillibaud Right, I think I understand. You don't want to check in `"dependency": "file:path/to/dependency"`. That's the one thing I agree is an added nuisance, but there are ways around it like I described above with my little script to swap them out at build/publish time.

What I wish they had done instead is allow us to create some kind of `local-dependencies.json` file and store those symlink references in there instead of `package.json`. And have `npm install` respect that file as an override for dependencies listed in `package.json`. Then you could just add `local-dependencies.json` to `.gitignore` and not have to modify `package.json` at all. That'd be the best of both worlds. You could totally delete `node_modues` and run `npm install` fresh locally and your symlinks would be restored (something you couldn't even do in npm 2 without having to redo all your links) and `package.json` wouldn't have to have hard references that need changed before being committed or published to production.

**Comment by dcaillibaud** [comment]
@chevex It would be great if npm could manage local overrides through e.g. `--local-overrides package.local.json`

Other solution could be to pass the package.json content to the preinstall script (and be aware of changes), because I tried to modify package.json file at preinstall runtime, but npm doesn't re-read the file before installing.

So, I tried your solution with this [script](https://github.com/npm/npm/files/1114971/npmi.txt) (github don't take js extension, so I put txt), but the syntax `"myModule": "file:../myModule" seems to copy files, not link directory !

This [proposition](https://github.com/npm/npm/pull/15900) could solve the pb, but in a partial way, still need a script in replacement of `npm install`, or one of the above suggestion to put this script in preinstall.

**Comment by dcaillibaud** [comment]
I made a script which replace `npm install`
- merge package.local-extras.json into package.json, stripping all "file:" dependencies (I don't want a copy)
- run `npm install`
- link all local "file:" dependencies
- restore original package.json

Use it at your own risk !
https://github.com/Sesamath/sesatheque/blob/0.9.74/scripts/npmi.js

**Comment by BeekeeperAlex** [comment]
You say it copied the files?  You sure you were using npm 5?  For me that file prefix always creates a Symlink  to the directory.

**Comment by dcaillibaud** [comment]
I saw this with npm@5.0.3, I'm pretty sure to have seen plain directories after a `rm -rf node_modules; npm i` (with some file: dependencies), but I can't reproduce it.

I just upgraded to 5.0.4, and I don't really understand, after erasing ~/.npm, putting my file:… in package.json, `npm install` doesn't install any of the file:… dependencies. If y specify explicitly one of my own module with `npm install myPackage`, the symlink is created. Then a `rm -rf node_modules; npm i` install registry modules and myPackage is linked, but other packages with file: aren't ! Like if you need to first install once the local module alone to force npm realize that the path is a real one…
It's the same after a revert to 5.0.3

**Comment by Madd0g** [comment]
Just "upgraded" to the latest (5.0.4) npm and started having this issue where my npm-linked deps are disappearing after every npm install. What's even weirder is that one of the deps of the linked package is seemingly disappearing as well, even though it's a regular npm package.

So I have to do `npm install` in the linked package and `npm link my-package` after every time npm does anything. OHHHHH.

FUN TIMES>>

**Comment by mark007** [comment]
I'm also seeing npm prune deleting dependencies of linked modules. I have a feeling this is the cause of alot of other existing npm issues that are still open. It looks like link and prune are not compatible with eachother at the moment.

**Comment by shazron** [comment]
Some info in this Twitter thread between @indutny, @zkat, and @iarna  
https://twitter.com/indutny/status/883742222419623937

**Comment by ghost** [comment]
A quick comment on the stealing problem + local development with symlinks and particularly since this has been open for a long time:

1) Don't use npm3 or npm4.
2) Skip over npm5 for the moment because it has race conditions which will be a secondary blocking issue and use yarn for now, although it should be very easy to switch back later.
3) If you do not need to publish your modules publically the following will work for you, else you will need to consider how and when you commit or avoid committing package.json before a publish.
4) With npm install ../module or yarn link you can use file specifiers and relative symlinks (e.g. within node_modules, module => ../../module) to avoid the stealing issue but also have symlinked local development. And there should be significant caching gains as well because more of the modules are now locally referenced.
5) Solution order is clone everything (if you haven't or want to try elsewhere first), yarn link each package (dependency sequence should not matter), and then on a second pass re-link package.json modules for each top-level module (since we've seeded the top-levels this will now work)

I did this using the following as a first test:

.bashrc

MODULES="module1 module2 module3 module4"

function yarn-checkout { time echo $MODULES | tr ' ' '\n' | xargs --max-args=1 --max-procs=8 -I{} git clone git@bitbucket.org:$USER/{}.git;}

function yarn-loop { cd $DIR; for d in `ls | egrep -v '?lock'`; do echo "## $d"; cd $DIR/$d; eval $1 ; cd $DIR; done;}

function yarn-install { yarn-loop "yarn link";}

alias yarn-lnk='for f in `cat package.json|grep file:|sed "s/\"//g"|cut -d: -f1`; do yarn link $f; done'
function yarn-auto { yarn-loop "yarn-lnk";}

Upgrades:
npm install -g npm@latest
npm install -g yarn

Above:
#1: Checkout in parallel since you would not expect clone errors and probably not need to read any log output. Parallel was ~20s in our case on roughly 40 private modules as an 8-way.
#2 Just a helper function
#3 Does the initial install on each top-level module.
#4 Does the second pass for private/dependent modules which should work because private linking depth should be one level from the perspective of each top-level.

Your package.json files should now look something like:

{
  "author": "First Last <username@company.co>",
  "name": "dev-ops",
  "description": "none",
  "version": "0.0.1",
  "repository": {
    "type": "git",
    "url": " https://bitbucket.org/username/dev-ops.git"
  },
  "main": "lib/index.js",
  "engines": {
    "node": ">= 0.10.26"
  },
  "dependencies": {
    "async-class": "^0.4.1",
    "aws-sdk": "^2.3.11",
    "bluebird": "^3.3.5",
    "cluster-utils": "file:../cluster-utils",         <- file specifier, relative path
    "config-loader": "file:../config-loader",   <- file specifier, relative path
    "minimist": "^1.2.0",
    "moment": "^2.12.0",
    "prompt": "1.0.0",
    "q": "^1.4.1",
    "supervisor": "~0.6.0",
    "underscore": "^1.8.3"
  }
}

This worked well for us, moved us away from npm3 + the original npm link strategy, away from global symlinks, is now filesystem layout independent, and was also faster and/or simplifying in many cases in terms of dev-ops. I think the npm5 maintainers have thought about this problem, seems the solution is just not well understood.

Hope this helps.

**Comment by jflatow** [comment]
I too noticed this change in behavior, and wish `npm install` / `npm update` would ignore symlinks in `node_modules` as before.

**Comment by bertho-zero** [comment]
I went to yarn who corrected this problem today, npm knows problem since version 3! (https://github.com/yarnpkg/yarn/pull/4757)

**Comment by artus9033** [comment]
+1 for this still present issue 😞

**Comment by miensol** [comment]
We've experienced an error `npm ENOENT: no such file or directory, rename ...` this issue with local dependencies between modules inside single project.
E.g.
moduleA:
```
{
  "dependencies": {
    "aws-sdk": "^2.156.0"
  }
}
```

moduleB:
```
{
  "dependencies": {
    "jmespath": "0.15.0",
    "shared": "file:../moduleA"
  }
}
```

During the `npm --prefix moduleA install` all packages were installed including a dependency of `aws-sdk` shared called `"jmespath": "0.15.0"`. 

Now the `npm --prefix moduleB install` removed the `moduleA/node_modules/aws-sdk/node_modules/jmespath` folder and have put under `moduleB/node_modules/jmespath`. Now the `moduleA` `node_modules` is broken when looked at from the `moduleA` perspective. 

There are probably 2 issues here to be precise. `npm --prefix moduleB install` symlinked `moduleA` to `moduleB/node_modules/moduleA`. However the dependencies of `moduleA` seemed to not be installed. However, at the same time the command tried to reuse and flatten the dependencies of `moduleA`.

**Comment by UD-UD** [comment]
Hi, 
A bit late for the reply.
I have developed a cli utility that helps you to keep your local links unbroken even after installing dependencies.
You can download it from [npm](https://www.npmjs.com/package/npm-safe-install)

Also you can find the source code in [github](https://github.com/UD-UD/npm-safe-install)

Hope it helps. :)

**Comment by rdsedmundo** [comment]
NPM v6 still have this issue. Almost 3 years later. I'm wondering what's `big-bug` for the NPM team.

**Comment by zkat** [comment]
I want to note that one of our big major changes for the upcoming `npm@7` is going to be [an overhaul of how `npm link` works](https://github.com/npm/rfcs/pull/3), and this will probably be fully and thoroughly addressed in that release. For the time being, if anyone wants to keep this issue specifically alive, feel free to re-post it to [our new npm.community forum](https://npm.community), where we'll be moving all our issue tracking soon.

In the meantime, and because this thread is already long and impossible to actually consume, I'm going to close+lock this, because I don't intend to look in here again :)
